uc-crux-llvm/src/UCCrux/LLVM/Shape.hs

```diff
@@ -0,0 +1,424 @@
+      -- TODO print elements
+      fmap
+        ( PP.hsep
+            . ( [ PP.pretty "A struct:"
+                ]
+                  ++
+              )
+            . (: [])
+        )
+        (ppTag tag')
+
+ppShape ::
+  (forall ft'. tag ft' -> Doc Void) ->
+  Shape m tag ft ->
+  Doc Void
+ppShape ppTag = coerce (ppShapeA (Identity . ppTag))
+
+eqShape ::
+  (forall ft'. tag ft' -> tag ft' -> Bool) ->
+  Shape m tag ft ->
+  Shape m tag ft ->
+  Bool
+eqShape eqTag shape1 shape2 =
+  case (shape1, shape2) of
+    (ShapeInt tag1, ShapeInt tag2) -> eqTag tag1 tag2
+    (ShapeFloat tag1, ShapeFloat tag2) -> eqTag tag1 tag2
+    (ShapePtr tag1 ShapeUnallocated, ShapePtr tag2 ShapeUnallocated) -> eqTag tag1 tag2
+    (ShapePtr tag1 (ShapeAllocated n1), ShapePtr tag2 (ShapeAllocated n2)) ->
+      n1 == n2
+        && eqTag tag1 tag2 -- O(1)
+    (ShapePtr tag1 (ShapeInitialized rest1), ShapePtr tag2 (ShapeInitialized rest2)) ->
+      Vec.length rest1 == Vec.length rest2
+        && eqTag tag1 tag2 -- O(1)
+        && all (uncurry (eqShape eqTag)) (Vec.zip rest1 rest2)
+    (ShapeFuncPtr tag1, ShapeFuncPtr tag2) -> eqTag tag1 tag2
+    (ShapeOpaquePtr tag1, ShapeOpaquePtr tag2) -> eqTag tag1 tag2
+    (ShapeArray tag1 natRep1 rest1, ShapeArray tag2 natRep2 rest2) ->
+      natValue natRep1 == natValue natRep2
+        && eqTag tag1 tag2 -- O(1)
+        && and (PVec.zipWith (eqShape eqTag) rest1 rest2)
+    (ShapeStruct tag1 rest1, ShapeStruct tag2 rest2) ->
+      eqTag tag1 tag2
+        && allFC getConst (Ctx.zipWith (\s1 s2 -> Const (eqShape eqTag s1 s2)) rest1 rest2)
+    (_, _) -> False
+
+instance EqF tag => EqF (Shape m tag) where
+  eqF = eqShape eqF
+
+tag :: Simple Lens (Shape m tag ft) (tag ft)
+tag =
+  lens
+    ( \case
+        ShapeInt tg -> tg
+        ShapeFloat tg -> tg
+        ShapePtr tg _ -> tg
+        ShapeFuncPtr tg -> tg
+        ShapeOpaquePtr tg -> tg
+        ShapeArray tg _ _ -> tg
+        ShapeStruct tg _ -> tg
+    )
+    ( \s tg ->
+        ( case s of
+            ShapeInt _ -> ShapeInt tg
+            ShapeFloat _ -> ShapeFloat tg
+            ShapePtr _ rest -> ShapePtr tg rest
+            ShapeFuncPtr _ -> ShapeFuncPtr tg
+            ShapeOpaquePtr _ -> ShapeOpaquePtr tg
+            ShapeArray _ n rest -> ShapeArray tg n rest
+            ShapeStruct _ rest -> ShapeStruct tg rest
+        )
+    )
+
+data ShapeSeekError
+  = -- | First is index, second is size
+    NotEnoughElements !Int !Int
+  | DereferenceUnallocated
+  | DereferenceUninitialized
+
+ppShapeSeekError :: ShapeSeekError -> Text
+ppShapeSeekError =
+  Text.pack
+    . \case
+      NotEnoughElements _index _size -> "Not enough elements"
+      DereferenceUnallocated -> "Dereference of unallocated memory"
+      DereferenceUninitialized -> "Dereference of uninitialized memory"
+
+-- How is this not a thing?
+vUpdate :: Vec.Vector a -> Int -> a -> Vec.Vector a
+vUpdate vec idx a = Vec.update vec (Vec.singleton (idx, a))
```

Comment:
I think they designed the API that way to emphasize that it is expensive. You could use `ix` from `Control.Lens.At.ix`. `Data.Sequence` might be better if you anticipate a non-trivial number of updates

Reply:
Gotcha, thanks for the pointer. I do think `Sequence` is the better choice.

Reply:
Well, an in-place replace of `Data.Vector` with `Data.Sequence` causes `mutually_recursive_linked_list_sum.c` to start spinning. That's pretty confusing since the algorithmic complexity isn't changing in any *crazy* way:

- `Seq.length = Vec.length = O(1)`
- `Vec.! = O(1) < Seq.index = O(log(n))`
- `Seq.insertAt = O(log(n)) < vUpdate = O(n)`

Plus, overall the data in question should be pretty small for any of the tests given the low loop bounds. So I'm a bit at a loss for what could be happening. I could profile...?

Reply:
Commit available here: https://github.com/langston-barrett/crucible/commit/9e87af73e4fdd25f31a8e4b9fe90632d68e357f0

Reply:
I think you want `update` instead of `insertAt` - the latter shifts the rest of the sequence down, growing it

Reply:
:facepalm: That's right, and that fixed it! Thanks!

https://github.com/GaloisInc/crucible/pull/673/commits/2a52b09918670a327af0aa7738d107253dd014f8